### DISCUSSION TITLE
Shared topology in pyPrimeMesh

### DISCUSSION THREAD

**Discussion by lorenzovecchietti** [ansys/pyprimemesh#749] [Q&A] [answered]
Hi,
I'm trying to learn the basics of pyprimemesh. I wanted to do a share topology operation without resorting to spaceclaim/discovery.

I'm trying with a dummy example with a heatsink and a small enclosure around it.
So I have two parts (heatsink and enclosure) and I merge them --> I have two volume zones.

Now I want to apply the shared topology between the bottom of the heatsink and the bottom of the enclosure. Is it possible? How do I do this?

![immagine](https://github.com/ansys/pyprimemesh/assets/58366962/0758d308-c351-48b9-905a-c05f7bd0abea)

**Comment by agvarghe** [comment]
Hi @lorenzovecchietti 

PyPyrimeMesh do have an analogous feature similar to Share Topology available in SpaceClaim/ Discovery, Its called the scaffolder. Using scaffolder you can create a conformal interface between topologies. 

Here is a snippet you can use 
`
params = prime.ScaffolderParams(model=model,absolute_dist_tol=0.005,
                                                        intersection_control_mask=prime.IntersectionMask.FACEFACEANDEDGEEDGE,
                                                        constant_mesh_size=0.1,
                                                        )`
`
scaffolder = prime.Scaffolder(model, part.id)
res = scaffolder.scaffold_topo_faces_and_beams(
                                                                              topo_faces=part.get_topo_faces(), topo_beams=[], params=params
                                                                              )`
`
result=scaffolder.merge_overlapping_topo_faces(part.get_topo_faces(),params)`


Details regarding the api can be found in the link below 
PyPrimeMesh :  [Scaffolder API ](https://prime.docs.pyansys.com/version/stable/api/_autosummary/ansys.meshing.prime.Scaffolder.html)

**Reply by lorenzovecchietti** [reply]
After playing a bit with params values, I can obtain some effects, but the bottom surface of the enclosure is simply deleted:
![immagine](https://github.com/ansys/pyprimemesh/assets/58366962/4a172ab2-5e36-48fc-9689-f4e194a526d9)

vs the effect I was expecting
![immagine](https://github.com/ansys/pyprimemesh/assets/58366962/dc0bc578-83ca-4e7e-86be-9df3e0de4570)

**Comment by waltersma** [comment, marked as answer]
Hi @lorenzovecchietti , I would guess that the tolerances might be such that the face is being removed unintentionally.  I have prepared an example I hope will clarify things.  Please review and see if this helps.  

When doing the scaffolding, the step that is likely causing the issue here is that the resulting topo face has an interior edge loop rather than splitting the face in two.  If you then split this face you can delete one side of this and form the volumes as you expect.

```
import ansys.meshing.prime as prime
from ansys.meshing.prime.graphics import Graphics

cad_file = r"heatsink.fmd"
face_size = 2.0
tolerance = 0.04

with prime.launch_prime() as prime_client:
    model = prime_client.model
    prime.lucid.Mesh(model).read(cad_file)

    # merge parts to allow connection
    part_ids = [part.id for part in model.parts]
    params = prime.MergePartsParams(model)
    result = model.merge_parts(part_ids=part_ids, params=params)

    # scaffold topology to imprint adjacent faces
    # topo face imprinted with edge but not split
    # before:
    #    ----------------------------
    #          O---------------O
    #          |               |
    # 4 topo faces, 2 topo edges (not shared)
    #
    # after:
    #    ------O================O-----
    #          |                |
    # 4 topo faces, 2 topo edges (edges shared)
    print(model)
    part = model.get_part(result.merged_part_id)
    params = prime.ScaffolderParams(
        model,
        absolute_dist_tol=tolerance,
        intersection_control_mask=prime.IntersectionMask.FACEFACEANDEDGEEDGE,
        constant_mesh_size=face_size,
    )
    faces = part.get_topo_faces()
    beams = []
    result = prime.Scaffolder(model, part.id).scaffold_topo_faces_and_beams(
        topo_faces=faces,
        topo_beams=beams,
        params=params,
    )
    g = Graphics(model)
    print(model)
    g()

    # surface mesh with triangular elements of uniform size
    surfer_params = prime.SurferParams(
        model=model,
        size_field_type=prime.SizeFieldType.CONSTANT,
        constant_size=face_size,
    )
    surfer_result = prime.Surfer(model).mesh_topo_faces(
        part.id, topo_faces=part.get_topo_faces(), params=surfer_params
    )

    # split imprinted topo face by mesh regions
    # before:
    #    ------O================O-----
    #          |                |
    # 4 topo faces, 2 topo edges (edges shared)
    #
    # after:
    #    ------O≃≃≃≃≃≃≃≃≃≃≃O-----
    #          |                |
    # 5 topo faces, 2 topo edges (edges shared)
    prime.Scaffolder(model, part.id).split_topo_faces_by_mesh_region(
        part.get_topo_faces()
    )
    print(model)

    # remove duplicate topo faces
    # before:
    #    ------O≃≃≃≃≃≃≃≃≃≃≃O-----
    #          |                |
    # 5 topo faces, 2 topo edges (edges shared)
    #
    # after:
    #    ------O~~~~~~~~~~~~~~~~O-----
    #          |                |
    # 4 topo faces, 2 topo edges (edges shared)
    vol_params = prime.VolumetricScaffolderParams(
        model, absolute_dist_tol=tolerance, only_check_exact_overlaps=True
    )
    prime.Scaffolder(model, part.id).delete_shadowed_topo_faces(
        part.get_topo_faces(), vol_params
    )
    g = Graphics(model)
    print(model)
    g()

    # compute the newly modified topo volumes
    part.compute_topo_volumes(prime.ComputeVolumesParams(model))

    # volume mesh
    automesh_params = prime.AutoMeshParams(
        model=model,
        size_field_type=prime.SizeFieldType.CONSTANT,
    )
    prime.AutoMesh(model).mesh(part_id=part.id, automesh_params=automesh_params)

```
Cc. @agvarghe 
[heatsink.zip](https://github.com/ansys/pyprimemesh/files/15294656/heatsink.zip)

**Reply by lorenzovecchietti** [reply]
Hi @waltersma, sorry for the late reply; I didn't have the time to look into this further until today. Your approach works perfectly, thank you! :)
